### ISSUE TITLE
C semaphore wait and post hangs 

### ISSUE THREAD

**Issue by miniskar** [JuliaLang/julia#55635]
Dear Julia,
My C++ application is using semaphore and it is called from Julia. However, it hangs after semaphore wait (sem_wait) is called when there is sleep() function. If I remove the sleep() function call from the code, it hangs during the wait(task1). 

1. Output of versioninfo()
```julia-repl
julia> versioninfo()
Julia Version 1.10.5
Commit 6f3fdf7b362 (2024-08-27 14:19 UTC)
Build Info:
  Official https://julialang.org/ release
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: 64 × AMD Ryzen Threadripper 3970X 32-Core Processor
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-15.0.7 (ORCJIT, znver2)
Threads: 1 default, 0 interactive, 1 GC (on 64 virtual cores)
Environment:
  JULIA_VERSION = 1.10.5
  LD_LIBRARY_PATH = /home/nqx/packages/julia-1.10.5/lib:/home/nqx/install/lib:/home/nqx/install/lib:

```
2. How you installed Julia:
I have installed by downloading the Linux x86 (64) binary
3. MWE
c_code.c:
```c
#include <stdio.h>
#include <pthread.h>
#include <semaphore.h>

sem_t sem;
void c_sem_init() {
    sem_init(&sem, 0, 0);  // Initialize semaphore with a value of 0
}
void c_function_with_wait() {
    int res = sem_wait(&sem);
    if (res != 0) {
        perror("sem_wait failed");
    }
}
void c_function_with_post() {
    int res = sem_post(&sem);
    if (res != 0) {
        perror("sem_post failed");
    }
}
``` 

julia_c_code.jl:
```julia
# Define the C function signature
using Base.Threads;

function wait_on_semaphore()
    ccall((:c_function_with_wait, "libinfinite.so"), Cvoid, ())
end

function post_to_semaphore()
    ccall((:c_function_with_post, "libinfinite.so"), Cvoid, ())
end

gc_state = @ccall(jl_gc_safe_enter()::Int8)
ccall((:c_sem_init, "libinfinite.so"), Cvoid, ())
# Create tasks that call the respective functions
task1 = @spawn wait_on_semaphore()
println(Core.stdout, "Waiting on semaphore")
sleep(1)  # Ensure task1 runs and waits on the semaphore

println(Core.stdout, "Now posting on semaphore")
task2 = @spawn post_to_semaphore()
println(Core.stdout, "Posted on semaphore")

# Wait for both tasks to complete
wait(task1)
wait(task2)
@ccall(jl_gc_safe_leave(gc_state::Int8)::Cvoid)
```

Makefile:
```make
build:
    gcc -g -O0 -fPIC -shared  -o libinfinite.so c_code.c -I$(JULIA)/include/julia -L$(JULIA)/lib -ljulia -lpthread
run:
    julia julia_c_code.jl
```

Output:
```
Waiting on semaphore
```

Please help to resolve this issue.

**Comment by Seelengrab** [comment]
Welcome! This issue tracker is for bugs in the julia language, please consider posting to discourse.julialang.org instead for problems related to your program.

That being said, I believe the issue here is that you're running julia with one thread:

```
Threads: 1 default, 0 interactive, 1 GC (on 64 virtual cores)
```

Since the thread is blocked in the ccall, it will never yield back to the julia scheduler, which means that the julia scheduler can't schedule your main task (where you called `sleep` on) again.

**Comment by gbaraldi** [comment]
In addition blocking thread 0 outside of the scheduler is also very likely to cause a deadlock. (Thread 0 needs to run to advance any tasks waiting on libuv, which is likely with file operations or printing).

**Comment by miniskar** [comment]
Thank you for your responses and making me to understand the issue. I found a way to get around to avoid this issue by setting JULIA_NUM_THREADS environment variable. 

However, this is a limitation in Julia. Ideally there is no deadlock situation when you run this similar code in C++. 
In this scenario, we have asked to create two threads one thread is waiting on the semaphore and another thread is posting semaphore. The main julia code (thread) is calling sleep just before creating post thread. After creating post thread, it waits for the two threads to complete. Even if we remove sleep statement, it still hangs. The execution behavior of Julia scheduler for threads is creating the deadlock situation. It looks like thread switching is not happening when some thread is waiting for IO/sleep/semaphore. I am wondering whether Julia doesn't have thread switching?

**Comment by Seelengrab** [comment]
> In this scenario, we have asked to create two threads one thread is waiting on the semaphore and another thread is posting semaphore.

I think there is a misunderstanding here (I blame the name of the standard library) - `@spawn` _does not_ create new OS threads, it spawns a julia `Task` that is then executed on the available OS threads that julia was spawned with. These are sometimes called green threads, or an n:m threading model (n tasks scheduled on m OS threads).

>Even if we remove sleep statement, it still hangs. The execution behavior of Julia scheduler for threads is creating the deadlock situation

Yes, as mentioned above, the `sem_wait` in your C code blocks the sole available OS thread (the one the scheduler itself also runs on!) and doesn't yield to the julia scheduler. Because of this, your julia code doesn't get scheduled again and nothing ends up signaling the semaphore, resulting in the deadlock.

In other words, there's nothing the scheduler can do about this, it doesn't even get to run again.
